### PR TITLE
Enable `strict` mode

### PR DESCRIPTION
**Pull request by dstaley**
🎟️ [Asana Task](https://app.asana.com/0/1100423001970639/1201500876162997/f)
🔍 [Preview Link](https://react-components-git-{branch-slug}-hashicorp.vercel.app)

---

## Description

This PR enables `strict` in our `tsconfig`, in addition to fixing the compiler errors that arise as a result of the change. The only `strict` option we currently cannot enable is `noImplicitAny`, so that's been disabled.

For the most part, I preferred asserting that the current usage was valid rather than introduce a significant number of logic changes across several packages. In some instances, however, I felt it relatively safe to modify the logic to be valid with the types as written.

One of the largest set of changes came from our `ref` usage. The types for refs are not exactly the easiest things to work with, so if something looks funky please let me know and we can take a closer look. Everything passes the compiler, but that doesn't necessarily mean it's "correct"!

## Todo

- [ ] Add typechecking to CI
  - [ ] Do we want GitHub Actions or CircleCI?
  - [ ] Do we want to require a passing typecheck in order to commit to `main`?

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [ ] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [x] Add or update tests as appropriate.
- [ ] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [ ] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [ ] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

**Comment by BRKalow** [comment]
This seems like a great baseline to me. Acknowledging that usage of `!` _can_ be unsafe, but that everything is functionally unchanged with this, I think that's okay. Your ref typings look correct to me.

> Do we want GitHub Actions or CircleCI?

GitHub Actions!

> Do we want to require a passing typecheck in order to commit to main?

I would say yes!

File: packages/tabs/partials/tab-triggers/index.tsx [hunk 1]
```diff
@@ -61,7 +63,7 @@ function TabTriggers({
    */
   useEffect(() => {
     // Determine which arrows to show
-    const { scrollLeft, scrollWidth, offsetWidth } = scrollRef.current
+    const { scrollLeft, scrollWidth, offsetWidth } = scrollRef.current!
```
Comment (BRKalow) [inline packages/tabs/partials/tab-triggers/index.tsx:66]: This potentially feels like a scenario where we might want to check that `scrollRef.current` is defined, but as it was like this before I don't think it's critical.

File: packages/docs-page/server/index.ts [hunk 2]
```diff
@@ -117,7 +117,7 @@ export function generateStaticPaths({
   const loader = new FileSystemLoader({
     navDataFile,
     localContentDir,
-    product: product?.slug,
+    product: product?.slug ?? '',
```
Comment (thiskevinwang) [inline packages/docs-page/server/index.ts:120]:
```suggestion
    product: product.slug,
```

Comment (thiskevinwang) [reply packages/docs-page/server/index.ts:120]: And, I think [product isn't supposed to be optional](https://github.com/hashicorp/react-components/pull/481/files#diff-99829c81bc27331948908774fd93aa5a7e75b42d97274484f95d2c11d133025fR102) 🤔 

@BRKalow?

**Review by thiskevinwang — COMMENTED** [review]
Just 1 comment; And I'm a big fan of strict null checks! 🌶️ 🎉

File: packages/docs-page/server/index.ts (hunk 2 above)
Comment (BRKalow) [reply packages/docs-page/server/index.ts:120]: Right, it probably shouldn't be optional as we need it for the edit this page link.

Comment (dstaley) [reply packages/docs-page/server/index.ts]: 👍

**Review by BRKalow — APPROVED** [review]

**Pull request merged**